pkg/controller/migration/convert/network.go

```diff
@@ -0,0 +1,165 @@
+		hp := v1.HostPortsDisabled
+		install.Spec.CalicoNetwork.HostPorts = &hp
+	}
+
+	// CNI bandwidth plugin
+	if _, ok := c.pluginCNIConfig["bandwidth"]; ok {
+		return ErrIncompatibleCluster{"operator does not yet support bandwidth"}
+	}
+
+	if c.calicoCNIConfig == nil {
+		// TODO: don't return an error once we support this, instead just returning nil.
+		return ErrIncompatibleCluster{"operator does not yet support running without calico CNI"}
+	}
+
+	if c.calicoCNIConfig.MTU == -1 {
+		// if MTU is -1, we assume it was us who replaced it when doing initial CNI
+		// config loading. We need to pull it from the correct source
+		mtu, err := c.node.getEnv(ctx, c.client, containerInstallCNI, "CNI_MTU")
+		if err != nil {
+			return err
+		}
+		if mtu == nil {
+			return fmt.Errorf("couldn't detect MTU")
+		}
+
+		// TODO: clean this up
+		i := intstr.FromString(*mtu)
+		iv := int32(i.IntValue())
+		install.Spec.CalicoNetwork.MTU = &iv
+	} else {
+		// user must have hardcoded their CNI instead of using our cni templating engine.
+		// use the hardcoded value.
+		mtu := int32(c.calicoCNIConfig.MTU)
+		install.Spec.CalicoNetwork.MTU = &mtu
+	}
+
+	// CNI IPAM pools
+	if len(c.calicoCNIConfig.IPAM.IPv4Pools) != 0 {
+		return ErrIncompatibleCluster{"cni ipam ranges not suported"}
+	}
+
+	// Other CNI features
+	if c.calicoCNIConfig.FeatureControl.FloatingIPs {
+		return ErrIncompatibleCluster{"floating IPs not supported"}
+	}
+	if c.calicoCNIConfig.FeatureControl.IPAddrsNoIpam {
+		return ErrIncompatibleCluster{"IpAddrsNoIpam not supported"}
+	}
+
+	return nil
+}
+
+// getAutoDetection auto-detects the IP and Network using the requested
+// detection method.
+func getAutoDetection(method string) (operatorv1.NodeAddressAutodetection, error) {
+	const (
+		AUTODETECTION_METHOD_FIRST          = "first-found"
+		AUTODETECTION_METHOD_CAN_REACH      = "can-reach="
+		AUTODETECTION_METHOD_INTERFACE      = "interface="
+		AUTODETECTION_METHOD_SKIP_INTERFACE = "skip-interface="
+	)
+
+	// first-found
+	if method == "" || method == AUTODETECTION_METHOD_FIRST {
+		var t = true
+		return operatorv1.NodeAddressAutodetection{FirstFound: &t}, nil
+	}
+
+	// interface
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_INTERFACE) {
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_INTERFACE)
+		return operatorv1.NodeAddressAutodetection{Interface: ifStr}, nil
+	}
+
+	// can-reach
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_CAN_REACH) {
+		dest := strings.TrimPrefix(method, AUTODETECTION_METHOD_CAN_REACH)
+		return operatorv1.NodeAddressAutodetection{CanReach: dest}, nil
+	}
+
+	// skip-interface
+	if strings.HasPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE) {
+		ifStr := strings.TrimPrefix(method, AUTODETECTION_METHOD_SKIP_INTERFACE)
+		return operatorv1.NodeAddressAutodetection{SkipInterface: ifStr}, nil
+	}
+
+	return operatorv1.NodeAddressAutodetection{}, errors.New("unrecognized option for AUTODETECTION_METHOD_SKIP_INTERFACE: " + method)
```

Comment:
AUTODETECTION_METHOD_SKIP_INTERFACE isn't the right thing to report error on. I wonder if this function should return an error that the higher level function should add context to (because this function can be used for IPv6 or IPv4.

Reply:
Good point. I adjusted the message and wrapped that context from the caller :+1: